### PR TITLE
Enable warnings

### PR DESCRIPTION
**Pull request by sellout**
This turns on a lot of the “easy” warnings. I.e., things that are generally trivial to fix. It excludes `-Wsign-compare` and `-Wconversion`, which are very useful (and requested in #1266), but generally require more analysis to propagate changes through the code (or explicitly cast when we mean to).

This allows us to start ratcheting down on cases while the deeper ones are given more consideration.

### CODE REVIEW

File: src/gtest/test_history.cpp [hunk 1]
```diff
@@ -43,10 +43,14 @@ class FakeCoinsViewDB : public CCoinsView {
...
                     const uint256 &hashSaplingAnchor,
+                    const uint256 &hashOrchardAnchor,
                     CAnchorsSproutMap &mapSproutAnchors,
                     CAnchorsSaplingMap &mapSaplingAnchors,
+                    CAnchorsOrchardMap &mapOrchardAnchors,
                     CNullifiersMap &mapSproutNullifiers,
-                    CNullifiersMap saplingNullifiersMap) {
+                    CNullifiersMap &mapSaplingNullifiers,
+                    CNullifiersMap &mapOrchardNullifiers,
+                    CHistoryCacheMap &historyCacheMap) {
```
Comment (nuttycom) [inline src/gtest/test_history.cpp:53]: This is so frustrating. This is why I really want to get https://github.com/zcash/zcash/pull/5244 merged.

File: src/bench/crypto_hash.cpp [hunk 2]
```diff
@@ -60,7 +60,10 @@ static void SHA512(benchmark::State& state)
 static void FastRandom_32bit(benchmark::State& state)
 {
     FastRandomContext rng(true);
+    _Pragma("GCC diagnostic push")
+    _Pragma("GCC diagnostic ignored \"-Wunused-but-set-variable\"")
     uint32_t x = 0;
+    _Pragma("GCC diagnostic pop")
```
Comment (daira) [inline src/bench/crypto_hash.cpp:66]:
```suggestion
    [[maybe_unused]] uint32_t x = 0;
```
(not sure whether this works, but if it does [it is standard C++17](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused)).

File: src/gtest/test_history.cpp (hunk 1 above)
Comment (sellout) [reply src/gtest/test_history.cpp:53]: Now that that’s merged, this convo is resolved 😆

File: src/bench/crypto_hash.cpp (hunk 2 above)
Comment (sellout) [reply src/bench/crypto_hash.cpp]: Yeah, the `_Pragma` (at least in this first pass) are just placeholders. I’ve eliminated them all now. In this case, I went with returning the value that is calculated for the benchmark. But `[[maybe_unused]]` might work here, too. Happy to try it if you think it’s a better solution. (I used `[[maybe_unused]]` in other cases, like ranged `for` loops, where you can’t omit the variable name to prove that it’s unused.)

**Comment by sellout** [comment]
This just “fixes” all of the enabled warning cases, I’m happy to take different approaches where it makes sense, in basically this descending order of preference:
- a different fix for the warning (sometimes this needs a comment)
- silencing a single instance of the warning we don’t want to change (this should involve adding a comment to the silenced warning
- silencing the warnings across an entire file with `diagnostic push` at the top and `diagnostic pop` at the bottom (e.g., to avoid conflicty backports)
- silencing the warnings across an entire file or group of files with `diagnostic push`/`pop` around the import(s) (e.g., to avoid modifying subtree dependencies)

**Comment by sellout** [comment]
Superseded by #6636.

**Pull request closed without merging**